frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSquashAction.java

```diff
@@ -0,0 +1,171 @@
+package com.virtuslab.gitmachete.frontend.actions.base;
+
+import static com.virtuslab.gitmachete.backend.api.SyncToParentStatus.InSyncButForkPointOff;
+import static com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle.getNonHtmlString;
+import static com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle.getString;
+
+import com.intellij.notification.NotificationType;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.Presentation;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vcs.VcsNotifier;
+import com.intellij.vcs.log.VcsCommitMetadata;
+import git4idea.rebase.log.squash.GitSquashOperation;
+import git4idea.repo.GitRepository;
+import io.vavr.Tuple2;
+import io.vavr.collection.List;
+import io.vavr.control.Option;
+import kotlin.Unit;
+import kr.pe.kwonnam.slf4jlambda.LambdaLogger;
+import lombok.CustomLog;
+import lombok.experimental.ExtensionMethod;
+import lombok.val;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+
+import com.virtuslab.gitmachete.backend.api.ICommitOfManagedBranch;
+import com.virtuslab.gitmachete.frontend.actions.common.VcsCommitMetadataAdapterForSquash;
+import com.virtuslab.gitmachete.frontend.actions.contextmenu.CheckoutSelectedBranchAction;
+import com.virtuslab.gitmachete.frontend.actions.dialogs.GitNewCommitMessageActionDialog;
+import com.virtuslab.gitmachete.frontend.defs.ActionPlaces;
+import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
+import com.virtuslab.qual.guieffect.UIThreadUnsafe;
+
+@ExtensionMethod(GitMacheteBundle.class)
+@CustomLog
+public abstract class BaseSquashAction extends BaseGitMacheteRepositoryReadyAction
+    implements
+      IBranchNameProvider {
+
+  private final String NL = System.lineSeparator();
+
+  @Override
+  public LambdaLogger log() {
+    return LOG;
+  }
+
+  @Override
+  @UIEffect
+  protected void onUpdate(AnActionEvent anActionEvent) {
+    super.onUpdate(anActionEvent);
+
+    Presentation presentation = anActionEvent.getPresentation();
+
+    val branchNameOption = getNameOfBranchUnderAction(anActionEvent);
+    val nonRootBranchOption = branchNameOption.flatMap(bn -> getManagedBranchByName(anActionEvent, bn))
+        .filter(b -> b.isNonRoot())
+        .map(b -> {
+          assert b.isNonRoot() : "branch is root";
+          return b.asNonRoot();
+        });
```

Comment:
Nit: one idea how to get rid of `assert`:

```suggestion
        .flatMap(b -> b.isNonRoot() ? List.of(b.asNonRoot()) : List.empty())
```

(modulo method names, I don't remember Vavr naming)